### PR TITLE
[DPE-2289] [DPE-2388] Upgrade from 14/stable and add integration tests

### PR DESCRIPTION
**Pull request by marceloneppel**
## Issue
We cannot upgrade from `14/stable`.

There are no integration tests for the upgrade process.

## Solution
Add logic to correctly upgrade from `14/stable`.

The monitoring user and its credentials are being created when upgrading from `14/stable` (which doesn't have that user).

This also fixes the upgrade from a single unit cluster and updates the workload version.

Add tests that check if the upgrade succeeds, by using some assertions related to the upgrade lib and ensure that no writes from an application are lost during the upgrade process.

### CODE REVIEW

File: src/upgrade.py [hunk 1]
```diff
@@ -90,6 +94,23 @@ def _on_postgresql_pebble_ready(self, event: WorkloadEvent) -> None:
...
+            self._set_up_new_credentials_for_legacy()
+
+        try:
+            self.charm.unit.set_workload_version(
+                self.charm._patroni.rock_postgresql_version or "unset"
+            )
+        except PostgreSQLGetPostgreSQLVersionError:
+            # Don't fail on this, just log it.
+            logger.warning("Failed to get PostgreSQL version")
+
```
Comment (dragomirp) [inline src/upgrade.py:113]: Wouldn't the version be set anyway by the pebble_ready in `charm.py`?

File: src/upgrade.py [hunk 2]
```diff
@@ -120,6 +141,25 @@ def _on_upgrade_changed(self, _) -> None:
...
+    def _on_upgrade_charm_check_legacy(self, event: UpgradeCharmEvent) -> None:
+        if not self.peer_relation:
+            logger.debug("Wait all units join the upgrade relation")
+            return
+
+        if self.state:
+            # Do nothing - if state set, upgrade is supported
+            return
+
+        logger.warning("Upgrading from unsupported version")
```
Comment (dragomirp) [inline src/upgrade.py:153]:
```suggestion
        logger.warning("Upgrading from unspecified version")
```
Isn't this for current stable?

**Review by dragomirp — APPROVED** [review]

File: tests/integration/ha_tests/test_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            logger.info("Run pre-upgrade-check action")
+            action = await leader_unit.run_action("pre-upgrade-check")
+            await action.wait()
+
+            # Ensure the primary has changed to the first unit.
+            primary_name = await get_primary(ops_test, DATABASE_APP_NAME)
+            assert primary_name == f"{DATABASE_APP_NAME}/0"
+
+    local_charm = await ops_test.build_charm(".")
+    fault_charm = Path("/tmp/", local_charm.name)
```
Comment (dragomirp) [inline tests/integration/ha_tests/test_upgrade.py:176]: Fails in CI, we should most probably return a path here: https://github.com/canonical/postgresql-k8s-operator/blob/dpe-2289-upgrade-integration-tests/tests/integration/conftest.py#L74

File: src/charm.py [hunk 4]
```diff
@@ -541,21 +550,21 @@ def enable_disable_extensions(self, database: str = None) -> None:
         Args:
             database: optional database where to enable/disable the extension.
         """
-        orginial_status = self.unit.status
```
Comment (taurus-forever) [inline src/charm.py:544]: :-)

**Review by taurus-forever — APPROVED** [review]

File: src/upgrade.py (hunk 1 above)
Comment (marceloneppel) [reply src/upgrade.py]: Good catch. I fixed it on https://github.com/canonical/postgresql-k8s-operator/pull/210/commits/7c8bed563bf4ad55fb64ed316cf50ce26c1069f6.

File: src/upgrade.py (hunk 2 above)
Comment (marceloneppel) [reply src/upgrade.py]: Exactly. Fixed on https://github.com/canonical/postgresql-k8s-operator/pull/210/commits/7c8bed563bf4ad55fb64ed316cf50ce26c1069f6.

File: tests/integration/ha_tests/test_upgrade.py (hunk 3 above)
Comment (marceloneppel) [reply tests/integration/ha_tests/test_upgrade.py]: Good catch. I'm gonna revisit this later.

**Comment by taurus-forever** [comment]
@marceloneppel are we ready to merge this? I want to promote it to the `14/stable`. Tnx!

**Comment by marceloneppel** [comment]
> @marceloneppel are we ready to merge this? I want to promote it to the `14/stable`. Tnx!

Merging now.

**Pull request merged**